scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,162 @@
+  }
+
+  private def collectUsedVariables(anonClass: ScExtendsBlock): Array[VariableData] = {
+    var res: List[ScParameter] = Nil
+    anonClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = PsiTreeUtil.getParentOfType(p, classOf[ScClass])
+            val containingObject = PsiTreeUtil.getParentOfType(p, classOf[ScObject])
+            val isTopLevelScala3Definition = containingObject == null && containingClass == null
+            if (PsiTreeUtil.isAncestor(containingClass, anonClass, true) || PsiTreeUtil.isAncestor(containingObject, anonClass, true) || isTopLevelScala3Definition) {
+              res = p :: res
+            }
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+    res.reverse.distinct.map(parameter =>
+      new ScalaVariableData(parameter, true, parameter.`type`().getOrNothing)
+    ).toArray
+
+  }
+
+  private case class DialogResult(className: String, variables: Array[VariableData])
+
+  private def showRefactoringDialog(project: Project, anonClass: ScExtendsBlock, usedVariables: Array[VariableData], target: Either[ScFile, ScTemplateDefinition]) = {
+    val dialog = new ScalaAnonymousToInnerDialog(project, anonClass, usedVariables, target)
+    if (dialog.showAndGet()) {
+      Some(DialogResult(dialog.getClassName, dialog.getVariables))
+    } else
+      None
+  }
+
+  private def createClass(name: String, anonClass: ScExtendsBlock, variables: Array[VariableData], project: Project): ScClass = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val parametersText = variables.map(v => s"${v.name}: ${v.`type`.getPresentableText}").mkString(", ")
+    val parameterOldNameToNewName = variables.map(v => v.variable.getName -> v.name).toMap
+
+    val text = s"class $name($parametersText) extends ${anonClass.getText}"
+
+    val newClass = createElementFromText[ScClass](text, anonClass)
+
+    newClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            expression.handleElementRename(parameterOldNameToNewName(p.getName))
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+
+    newClass
+  }
+
+  @tailrec
+  private def findAnonymousClass(element: PsiElement): Option[ScExtendsBlock] =
+    Option(element) match {
+      case Some(n: ScNewTemplateDefinition) =>
+        n.findFirstChildByTypeScala[ScExtendsBlock](ScalaElementType.EXTENDS_BLOCK)
+      case Some(_) => findAnonymousClass(element.getParent)
+      case None => None
+    }
+
+  @tailrec
```

Comment:
there is no `parents` collection on `PsiElement`. if there were, I would have used that instead.

Reply:
It's available in the same extension methods wrapper as `parentOfType`:
`import org.jetbrains.plugins.scala.extensions.PsiElementExt`
Also, IntelliJ suggests to import it (though you need to first type the name)

<img width="500" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/7326e748-715c-49ae-9d15-99ca72cbf91c">

Reply:
> why wouldn't there be an ScFile at the top?

There should be.
But it's generally not nice to use `.get`, so with the explicit comment I reassure that we don't expect NPE because at least `ScFile` is expected as one of the parents.

With your implementation, I would add a similar comment to `elem.getParent`